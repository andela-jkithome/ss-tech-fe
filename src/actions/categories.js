import fetch from 'isomorphic-fetch';

export const fetchRequest = () => ({
  type: 'CATEGORIES_REQUEST',
});

export const fetchSuccess = categories => ({
  type: 'CATEGORIES_SUCCESS',
  categories,
});

export const fetchFailure = failure => ({
  type: 'CATEGORIES_FAILURE',
  failure,
});

export const fetchCategories = url => dispatch => {
  dispatch(fetchRequest());
  return fetch(url)
    .then(response => {
      if (!response.ok) {
        return response.json().then(Promise.reject.bind(Promise));
      }
      return response.json();
    })
    .then(json => dispatch(fetchSuccess(json)))
    .catch(() => dispatch(fetchFailure('Error Fetching Categories!')));
};
