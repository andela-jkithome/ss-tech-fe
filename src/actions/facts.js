import fetch from 'isomorphic-fetch';

export const fetchRequest = () => ({
  type: 'FACT_REQUEST',
});

export const fetchSuccess = fact => ({
  type: 'FACT_SUCCESS',
  fact,
});

export const fetchFailure = failure => ({
  type: 'FACT_FAILURE',
  failure,
});

export const fetchFact = url => dispatch => {
  dispatch(fetchRequest());
  return fetch(url)
    .then(response => {
      if (!response.ok) {
        return response.json().then(Promise.reject.bind(Promise));
      }
      return response.json();
    })
    .then(json => dispatch(fetchSuccess(json)))
    .catch(() => dispatch(fetchFailure('Error Fetching Fact!')));
};
