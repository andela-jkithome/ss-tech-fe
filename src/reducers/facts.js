const fact = (
  state = {
    fetching: false,
    fact: {},
    error: null,
  },
  action,
) => {
  switch (action.type) {
    case 'FACT_REQUEST':
      return Object.assign({}, state, { fetching: true });
    case 'FACT_SUCCESS':
      return Object.assign({}, state, { fetching: false, fact: action.fact });
    case 'FACT_FAILURE':
      return Object.assign({}, state, {
        fetching: false,
        error: action.failure,
      });
    default:
      return state;
  }
};

export default fact;
