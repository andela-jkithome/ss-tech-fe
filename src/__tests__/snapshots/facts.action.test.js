import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';
import { fetchFact } from '../../actions/facts';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const host = 'https://api.chucknorris.io';
const url = 'https://api.chucknorris.io/jokes/random?category=science';

const fetchedFact = {
  category: ['science'],
  icon_url: 'https://assets.chucknorris.host/img/avatar/chuck-norris.png',
  id: 'uojt-t8as5ws5h1q-sjslw',
  url: 'https://api.chucknorris.io/jokes/uojt-t8as5ws5h1q-sjslw',
  value: `Nothing can escape the gravity of a black hole, except for Chuck Norris. 
  Chuck Norris eats black holes. They taste like chicken.`,
};

describe('async actions', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    nock.cleanAll();
  });

  it('creates FACT_SUCCESS when fetching joke has been done', () => {
    nock(host)
      .get('/jokes/random?category=science')
      .reply(200, fetchedFact);

    return store.dispatch(fetchFact(url)).then(() => {
      // return of async actions
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FACT_FAILURE when fetching joke returns an error', () => {
    nock(host)
      .get('/api/todos')
      .reply(500, { message: 'Error Fetching Fact' });

    return store.dispatch(fetchFact(url)).then(() => {
      // return of async actions
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
