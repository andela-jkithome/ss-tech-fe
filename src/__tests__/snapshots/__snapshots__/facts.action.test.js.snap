// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async actions creates FACT_FAILURE when fetching joke returns an error 1`] = `
Array [
  Object {
    "type": "FACT_REQUEST",
  },
  Object {
    "failure": "Error Fetching Fact!",
    "type": "FACT_FAILURE",
  },
]
`;

exports[`async actions creates FACT_SUCCESS when fetching joke has been done 1`] = `
Array [
  Object {
    "type": "FACT_REQUEST",
  },
  Object {
    "fact": Object {
      "category": Array [
        "science",
      ],
      "icon_url": "https://assets.chucknorris.host/img/avatar/chuck-norris.png",
      "id": "uojt-t8as5ws5h1q-sjslw",
      "url": "https://api.chucknorris.io/jokes/uojt-t8as5ws5h1q-sjslw",
      "value": "Nothing can escape the gravity of a black hole, except for Chuck Norris. 
  Chuck Norris eats black holes. They taste like chicken.",
    },
    "type": "FACT_SUCCESS",
  },
]
`;
